lib/dao.js

```diff
@@ -2632,6 +2632,8 @@ DataAccessObject.replaceById = function(id, data, options, cb) {
           'replaceById() must be implemented by the connector');
 
   var pkName = idName(this);
+  var initialData = {};
+  copyData(data, initialData);
```

Comment:
>Hmm, creating a copy of all data is rather expensive operation. Could you please take another look and check if there is a way how to avoid that?

Right, now I use a flag...

>Also consider the data passed to hooks, e.g. "before save" hook. I think the data should not include primary key when the user is not including it.

Sorry I'm a little bit confused here; for `before save` we are not passing `data` in context; please see [here](https://github.com/strongloop/loopback-datasource-juggler/blob/063983ccc9257db2e500eb273a585e6a9dad6652/lib/dao.js#L2662-L2670)

>It makes me wonder, perhaps we are hard-setting the primary key to prevent people from changing primary key via updateAttributes?

No for `updateAttributes` you can still pass the value of the PK in data object, but it should be the same the PK value for `instance` (it should pass `idEquals(data[idName], inst[idName])`); please see [here](https://github.com/strongloop/loopback-datasource-juggler/blob/063983ccc9257db2e500eb273a585e6a9dad6652/lib/dao.js#L2832-L2844)